ledger/narwhal/subdag/src/lib.rs

```diff
@@ -141,20 +172,20 @@ impl<N: Network> Subdag<N> {
         self.values().flatten().flat_map(BatchCertificate::transmission_ids)
     }
 
-    /// Returns the timestamp of the anchor round, defined as the median timestamp of the subdag.
-    pub fn timestamp(&self) -> i64 {
+    /// Returns the timestamp of the anchor round, defined as the weighted median timestamp of the subdag.
+    pub fn timestamp(&self, committee: &Committee<N>) -> i64 {
         match self.leader_certificate() {
             BatchCertificate::V1 { .. } => self.leader_certificate().timestamp(),
             BatchCertificate::V2 { .. } => {
-                // Retrieve the timestamps of the certificates.
-                let mut timestamps = self.values().flatten().map(BatchCertificate::timestamp).collect::<Vec<_>>();
-                // Sort the timestamps.
-                #[cfg(not(feature = "serial"))]
-                timestamps.par_sort_unstable();
-                #[cfg(feature = "serial")]
-                timestamps.sort_unstable();
-                // Return the median timestamp.
-                timestamps[timestamps.len() / 2]
+                // Retrieve the timestamps and stakes of the certificates.
+                let timestamps_and_stakes = self
+                    .values()
+                    .flatten()
+                    .map(|certificate| (certificate.timestamp(), committee.get_stake(certificate.author())))
```

Comment:
should we filter out the elements where the stake is equal to `0`? it would speed up sorting etc.

Reply:
In theory, there can't be a committee member with 0 stake. This may be a needless filter